### ISSUE TITLE
Additional Public Properties on Proxied object, needed ?

### ISSUE THREAD

**Issue by ordinaryorange** [mfidemraizer/trackerdog#10]
Got another one that cropped up today.

I notice that a proxy of a tracked object gets all the properties of the underlying object, but in the process of making the object Trackable, the proxy also gets 2 additional public properties. `ChangeTracker` and `CollectionProperties`
In my application I was throwing the underlying object into WPF PropertyGrid. Now when I pass the proxy instead of the underlying object I get to see the extra 2 public properties.

What are your thoughts about this ?
1. Should TrackerDog be adding public properties to a proxied object ? Or should the proxied object only republish those properties that are public in the underlying object.
2. Do they need to be public ? Will TrackerDog still work if they are internal ?
3. Or should the proxy be viewed as a subclass and therefore perfectly valid, and thus the responsibility of the consuming application to manage ?

**Comment by mfidemraizer** [comment]
Uhm now I can't give you an instant conclusion, but it should be a solution for this too.

**Comment by ordinaryorange** [comment]
Not looking for an instant solution. I'd rather you take time to consider any posts, that way the outcome will be better. ;) As normal have temporary workarounds for things in the mean time.

**Comment by mfidemraizer** [comment]
@ordinaryorange Nah, it was easy. Now both properties are private so you won't experience your described problem.

Try it out upgrading to NuGet package version 1.2.5! 

Let me know if you find it solved!

**Comment by ordinaryorange** [comment]
Hmm, just upgraded to 1.2.5, but both property getters still appear to be public.

**Comment by mfidemraizer** [comment]
@ordinaryorange Uhm, Castle Dynamic Proxy has turned explicitly-implemented interface properties into public properties... Now's time for a 2nd take!

**Comment by mfidemraizer** [comment]
@ordinaryorange They must be public because they're interface properties... I can't turn them into `internal`... I'll need to think any other alternative to solve this issue.

**Comment by mfidemraizer** [comment]
@ordinaryorange I finding hard to solve this issue in an elegant way.

`IChangeTrackableObject` is implemented by `ChangeTrackableObjectMixin` which is the class that gets mixed with yours to build the whole proxy. Since I assume that we both think that self-tracked objects is the way to go, I don't find any way of associating change tracking-specific data to a given proxy without implementing an interface. 

I'm still thinking about an alternative. Now our problem is Castle Dynamic Proxy doesn't mix interface explicitly-implemented members as *explicit* members in the resulting proxy....

**Comment by ordinaryorange** [comment]
There is the `BrowsableAttribute` ([https://msdn.microsoft.com/en-us/library/system.componentmodel.browsableattribute(v=vs.110).aspx]) which would solve my direct issue but that feels like a hack.

**Comment by ordinaryorange** [comment]
Is late here, will check on your progress in the morning.

**Comment by mfidemraizer** [comment]
@ordinaryorange Uhm it could be a possible solution, yes. It's not dangerous at all.

BTW, I've added a Q&A on StackOverflow: http://stackoverflow.com/questions/38593185/can-a-castle-dynamic-generated-proxy-be-forced-to-implement-members-as-explicit

Where are you from? Are you in Asia?

**Comment by mfidemraizer** [comment]
I've pushed a new NuGet package v1.2.6 with these attributes applied on the whole properties. Let me know if it solves your issue for now. 

BTW, I'll try to solve in a more elegant way, but the most important thing is that you get what you want from now!

**Comment by ordinaryorange** [comment]
Downloaded the latest update, but alas no good. I still see the two additional properties.
For now I've managed to implement a workaround. Fortunately I was using https://github.com/MatthewKing/DynamicDescriptors to create the object instance for the PropertyGrid, so I've added a temporary line in my code to inspect for the 2 unwanted properties and filter them out.
Let me know when you you come up with a more elegant solution.
PS. I live with the Kangaroos !

**Comment by mfidemraizer** [comment]
@ordinaryorange  It seems hard to workaround this unless you provide to WinForms/WPF/SL which properties you want to bind, am I mistaken?

I've filled an issue on Castle DynamicProxy project and I feel that our definitive solution is that this proxy generator could implement interfaces **explicitly** so their members would technicallly be **private** and you wouldn't experience the whole issue anymore...

Another solution could be moving the storage of both `ChangeTracker` or `CollectionProperties` to a *change tracking manager* which could be a dictionary where *keys* would be the proxies and the values a new class that could hold both properties, so `ChangeTrackableObjectMixin` wouldn't need to implement these public properties anymore...

I'll investigate further this last approach I've just invented in a few seconds, because we should agree that I'll be able to refactor TrackerDog in less time than convincing Castle's guys to implement the *implement interfaces explicitly* feature..............

**Comment by ordinaryorange** [comment]
Yes I checked your SO post and there was little hope there. I have a workaround for now so there is no time pressure from me. I like your ideas about moving the storage of the two properties to another class. This would keep the proxy to a more "pure" representation of the type is is wrapping.

**Comment by mfidemraizer** [comment]
@ordinaryorange Yup, and I believe that it won't be a big refactor, it'll be just a good separation of concerns ;)

**Comment by mfidemraizer** [comment]
I've already implemented this enhancement, but I've not pushed a NuGet package. Let's await for #7 issue to get implemented. Since you've found a temporal workaround, you should be able to wait for the next release :)

I'll update this issue soon.

**Comment by ordinaryorange** [comment]
nice work, yep can wait for #7 too.

**Comment by mfidemraizer** [comment]
Done! This fix is available on NuGet package version 2.0.0 and above ;)
